### DISCUSSION TITLE
Generate Empty Contstructor + Required Properties for Positional Record

### DISCUSSION THREAD

**Discussion by fbeauchemin** [dotnet/csharplang#7304] [Language Ideas] [answered]
With C# 11 `required` keyword, I propose that a _parameterless_ constructor be generated and that properties be marked as `required`. We could then use either the constructor or initializer approach when creating **positional record**:

```csharp
public record Example(int Prop1, string Prop2);

var r1 = new Example(10, "abc");
var r2 = new Example { Prop1 = 10, Prop2 = "abc" };
```

Currently, this is not possible as no _parameterless_ constructor exist. It would be equivalent to

```csharp
public record Example
{
    public Example() {}
    
    [SetsRequiredMembers]
    public Example(int Prop1, string Prop2)
    {
        this.Prop1 = Prop1;
        this.Prop2 = Prop2;
    }
    
    public required int Prop1 { get; init; }
    public required string Prop2 {get; init; }
}
```

with less code.

**Comment by CyrusNajmabadi** [comment]
It would be possible to easily accomplish this with a dedicated source generator that just emits the extra members you want.  We would not be likely to ever make this default as it would now add to the surface area of existing types in a way that many users might not want.

**Reply by fbeauchemin** [reply]
Not sure how it would impact existing types as the generated code currently implies `required` properties through the constructor. Guessing you mean having more than 1 constructor?

**Reply by CyrusNajmabadi** [reply, marked as answer]
You said that writing `public record Example(int Prop1, string Prop2);` would generate something equivalent to:

```c#
public record Example
{
    public Example() {}
    
    [SetsRequiredMembers]
    public Example(int Prop1, string Prop2)
    {
        this.Prop1 = Prop1;
        this.Prop2 = Prop2;
    }
    
    public required int Prop1 { get; init; }
    public required string Prop2 {get; init; }
}
```

Given that the initial code is already *legal* code with existing meaning, and an existing surface area generated, it would be bad for this to start giving a new surface area for existing users of records that do not expect it, or want it.  As such, a source-generator (which someone can opt into instead) could generate exactly that body for the type.

**Reply by fbeauchemin** [reply]
Thanks for the clarification!

**Comment by KennethHoff** [comment]
I never use the primary constructor-way of making records because of this. Constructors are terrible to work with for anything that has more than 1 value. I always use object initializers and required properties are practically mandatory for those.

I'd really love this change personally, as it's annoying having to write like 4x more code just because I want a single keyword on the underlying properties, but :shrug: 

Let's hope [this](https://github.com/dotnet/csharplang/issues/188) ever goes anywhere and maybe that could also allow for `required` as a primary ctor keyword. Even better would be to simply have `required record class`.

I practically always use required on everything nowadays. If I could make `required` and `sealed` default that'd be great.

**Reply by CyrusNajmabadi** [reply]
> as it's annoying having to write like 4x more code 

You can write a generator, and it can spit it all out for you.  No need to manually do anything.

**Reply by KennethHoff** [reply]
:thinking: I'll look into this..
